File: python/templative/lib/produce/customComponents/svgscissors/svgScissors.py
import os
from xml.etree import ElementTree
from aiofile import AIOFile
from svgmanip import Element
from lxml import etree

from templative.lib.componentInfo import COMPONENT_INFO
from templative.lib.manage.models.produceProperties import ProduceProperties, PreviewProperties
from templative.lib.manage.models.gamedata import PieceData, ComponentBackData
from templative.lib.produce.translation import getTranslation
from templative.lib.manage.models.composition import ComponentComposition
from templative.lib.manage.models.artdata import ComponentArtdata
from templative.lib.produce.customComponents.svgscissors.inkscapeProcessor import exportSvgToImage
from templative.lib.produce.customComponents.svgscissors.fontCache import FontCache

async def convertElementToString(element) -> str:
    out = etree.tostring(element.root, xml_declaration=True, standalone=True)
    return out.decode('utf-8')

async def createArtFileOfPiece(compositions: ComponentComposition, artdata: any, gamedata: PieceData | ComponentBackData, componentBackOutputDirectory: str, productionProperties: ProduceProperties | PreviewProperties, _:FontCache) -> None:
    templateFilesDirectory = compositions.gameCompose["artTemplatesDirectory"]
    if artdata is None: 
        print("!!! Missing artdata %s" % gamedata.componentDataBlob["displayName"])
        return
    artFilename = "%s.svg" % (artdata["templateFilename"])
    artFilepath = os.path.normpath(os.path.join(productionProperties.inputDirectoryPath, templateFilesDirectory, artFilename))
    if not os.path.exists(artFilepath):
        print("!!! Template art file %s does not exist." % artFilepath)
        return
    
    if not compositions.componentCompose["type"] in COMPONENT_INFO:
        raise Exception("No image size for %s", compositions.componentCompose["type"])
    component = COMPONENT_INFO[compositions.componentCompose["type"]]

    contents = None
    try:
        async with AIOFile(artFilepath, 'r') as f:
            contents = await f.read()
    except Exception as e:
        print(f"!!! Template art file {artFilepath} cannot be parsed. Error: {e}")
        return

    pieceName = gamedata.pieceData["name"] if isinstance(gamedata, PieceData) else gamedata.componentBackDataBlob["name"]
    imageSizePixels = component["DimensionsPixels"]
    
    contents = await addOverlays(contents, artdata["overlays"], compositions, gamedata, productionProperties)
    contents = await textReplaceInFile(contents, artdata["textReplacements"], gamedata, productionProperties)
    contents = await updateStylesInFile(contents, artdata["styleUpdates"], gamedata)
    contents = await addNewlines(contents)
    
    pieceUniqueHash = f"_{gamedata.pieceUniqueBackHash}" if gamedata.pieceUniqueBackHash != '' else ''
    artFileOutputName = f"{compositions.componentCompose['name']}{pieceUniqueHash}-{pieceName}"
    artFileOutputFilepath = await createArtfile(contents, artFileOutputName,imageSizePixels, componentBackOutputDirectory)
    await exportSvgToImage(artFileOutputFilepath, imageSizePixels, artFileOutputName, componentBackOutputDirectory)
    print(f"Produced {pieceName}.")

async def addNewlines(contents):
    if contents is None:
        raise Exception("contents cannot be None")
    return contents.replace("NEWLINE", "\n")

async def createArtfile(contents, artFileOutputName, imageSizePixels, outputDirectory):
    if contents is None:
        raise Exception("Contents cannot be None")

    root = ElementTree.fromstring(contents)
    
    SVG_NS = "http://www.w3.org/2000/svg"
    ElementTree.register_namespace('', SVG_NS)
    ElementTree.register_namespace('svg', SVG_NS)
    
    scale_factor = 96/300
    
    root.set('width', f'{imageSizePixels[0]*scale_factor}')
    root.set('height', f'{imageSizePixels[1]*scale_factor}')
    
    root.set('viewBox', f'0 0 {imageSizePixels[0]*scale_factor} {imageSizePixels[1]*scale_factor}')
    
    wrapper = ElementTree.Element(f'{{{SVG_NS}}}g')
    wrapper.set('transform', f'scale({scale_factor})')
    
    for child in list(root):
        wrapper.append(child)
    
    for child in list(root):
        root.remove(child)
    root.append(wrapper)
    
    contents = ElementTree.tostring(root, encoding='unicode')
    
    artFileOutputFileName = "%s.svg" % (artFileOutputName)
    artFileOutputFilepath = os.path.join(outputDirectory, artFileOutputFileName)
    async with AIOFile(artFileOutputFilepath, mode="w", encoding="utf-8") as file:
        await file.write(contents)
    return artFileOutputFilepath

async def addOverlays(contents, overlays, compositions: ComponentComposition, pieceGamedata: PieceData, productionProperties: ProduceProperties | PreviewProperties):
    if contents is None:
        raise Exception("contents cannot be None")

    if overlays is None:
        raise Exception("overlays cannot be None")

    overlayFilesDirectory = compositions.gameCompose["artInsertsDirectory"]

    for overlay in overlays:
        isComplex = overlay.get("isComplex", False)
        if isComplex and productionProperties.isSimple:
            continue

        isDebug = overlay.get("isDebugInfo", False)
        if isDebug and productionProperties.isPublish:
            continue

        positionX = overlay.get("positionX", 0)
        positionY = overlay.get("positionY", 0)
        overlayName = await getScopedValue(overlay, pieceGamedata)
        if not overlayName:
            continue

        overlaysFilepath = os.path.abspath(os.path.join(productionProperties.inputDirectoryPath, overlayFilesDirectory))
        overlayFilename = "%s.svg" % (overlayName)
        overlayFilepath = os.path.normpath(os.path.join(overlaysFilepath, overlayFilename))

        if not os.path.exists(overlayFilepath):
            print("!!! Overlay %s does not exist." % overlayFilepath)
            continue

        contents = await placeOverlay(contents, overlayFilepath, positionX, positionY)
    return contents

async def placeOverlay(contents:str, overlayFilepath, positionX, positionY) -> str:
    SVG_NS = "http://www.w3.org/2000/svg"
    ElementTree.register_namespace('', SVG_NS)
    ElementTree.register_namespace('svg', SVG_NS)
    
    main_svg = ElementTree.ElementTree(ElementTree.fromstring(contents))
    main_svg_root = main_svg.getroot()
    
    async with AIOFile(overlayFilepath, 'r') as afp:
        overlay_contents = await afp.read()
    overlay_svg_root = ElementTree.fromstring(overlay_contents)
    
    # Create group with proper namespace
    group = ElementTree.Element(f'{{{SVG_NS}}}g')
    
    # Instead of looking for the first transform group, add directly to root
    # and compensate for the root viewBox scaling
    viewBox = main_svg_root.get('viewBox', '').split()
    if viewBox:
        try:
            _, _, width, height = map(float, viewBox)
            actual_width = float(main_svg_root.get('width', width))
            scale_factor = actual_width / width
        except ValueError:
            scale_factor = 1.0
    else:
        scale_factor = 1.0
    
    # Apply position and inverse of root scaling
    group.set('transform', f'translate({positionX},{positionY}) scale({1/scale_factor})')
    
    # Copy attributes and children from overlay
    for key, value in overlay_svg_root.attrib.items():
        if key not in ['width', 'height', 'viewBox', 'transform']:
            group.set(key, value)
    
    for element in overlay_svg_root:
        group.append(element)
    
    # Add directly to root SVG instead of nested transform group
    main_svg_root.append(group)
    
    return ElementTree.tostring(main_svg_root, encoding='unicode')

async def textReplaceInFile(contents: str, textReplacements, gamedata: PieceData | ComponentBackData, productionProperties: ProduceProperties | PreviewProperties):
    for textReplacement in textReplacements:
        key = "{%s}" % textReplacement["key"]
        value = await getScopedValue(textReplacement, gamedata)
        if isinstance(value, str):
            value = value.replace("\\\\n", "\n")
        value = await processValueFilters(value, textReplacement)

        isComplex = textReplacement.get("isComplex", False)
        if isComplex and productionProperties.isSimple:
            value = ""

        isDebug = textReplacement.get("isDebugInfo", False)
        if isDebug and productionProperties.isPublish:
            value = ""

        if productionProperties.targetLanguage != "en" and textReplacement.get("isTranslateable", False) and value:
            translation = getTranslation("./", value, productionProperties.targetLanguage)
            if translation:
                value = translation
            else:
                print("Could not translate %s" % value)
        contents = contents.replace(key, value)
    return contents

async def processValueFilters(value, textReplacement):
    if "filters" in textReplacement:
        for filter in textReplacement["filters"]:
            if filter == "toUpper":
                value = value.upper()
    return str(value)

async def updateStylesInFile(contents, styleUpdates, pieceGamedata: PieceData):
    if contents is None:
        raise Exception("contents cannot be None.")

    if styleUpdates is None:
        raise Exception("styleUpdates cannot be None.")
    
    elementTree = ElementTree.ElementTree(ElementTree.fromstring(contents))
    try:
        for styleUpdate in styleUpdates:
            findById = styleUpdate["id"]
            elementToUpdate = elementTree.find(".//*[@id='%s']" % findById)
            if elementToUpdate is not None:
                value = await getScopedValue(styleUpdate, pieceGamedata)
                await replaceStyleAttributeForElement(elementToUpdate, "style", styleUpdate["cssValue"], value)
            else:
                print("Could not find element with id [%s]." % (findById))
        return ElementTree.tostring(elementTree.getroot(), encoding='unicode')
    except ElementTree.ParseError as pe:
        print("Production failed!", pe)

async def replaceStyleAttributeForElement(element:Element, attribute, key, value):
    attributeValue = element.get(attribute, "")
    replaceStyleWith = ""
    found = False

    cssKeyValuePairs = attributeValue.split(';')
    for cssKeyValuePair in cssKeyValuePairs:
        keyAndPair = cssKeyValuePair.split(':')
        if keyAndPair[0] == key:
            replaceStyleWith += f"{key}:{value};"
            found = True
        else:
            replaceStyleWith += cssKeyValuePair + ';'

    if not found:
        replaceStyleWith += f"{key}:{value};"

    if replaceStyleWith.endswith(";"):
        replaceStyleWith = replaceStyleWith[:-1]
    element.set(attribute, replaceStyleWith)

async def getScopedValue(scopedValue, pieceGameData: PieceData | ComponentBackData):
    if scopedValue is None:
        raise Exception("scopedValue cannot be None.")

    scope = scopedValue["scope"]
    source = scopedValue["source"]

    scopeData = None
    if scope == "studio":
        scopeData = pieceGameData.studioDataBlob
    elif scope == "game":
        scopeData = pieceGameData.gameDataBlob
    elif scope == "component":
        scopeData = pieceGameData.componentDataBlob
    elif scope == "piece":
        scopeData = pieceGameData.pieceData if isinstance(pieceGameData, PieceData) else pieceGameData.componentBackDataBlob
    elif scope == "global":
        return source
    elif scope == "utility":
        utilityFunctions = {
            # "git-sha": getCurrentGitSha
        }
        if source not in utilityFunctions:
            print("Missing function %s not found in %s scope." % (source, scope))
            return source
        return utilityFunctions[source]()

    if source not in scopeData:
        print("Missing key %s not found in %s scope." % (source, scope))
        return source

    return scopeData[source]

# def getCurrentGitSha():
#     repo = git.Repo(search_parent_directories=True)
#     return repo.head.object.hexsha

File: src/main/templative/lib/produce/customComponents/svgscissors/svgScissors.js
const os = require('os');
const path = require('path');
const fs = require('fs-extra');
const { parseStringPromise, Builder } = require('xml2js');
const { DOMParser, XMLSerializer } = require('@xmldom/xmldom');

const { COMPONENT_INFO } = require('../../../../../../shared/componentInfo.js');
const { ProduceProperties, PreviewProperties } = require('../../../manage/models/produceProperties.js');
const { PieceData, ComponentBackData } = require('../../../manage/models/gamedata.js');
const { getTranslation } = require('../../translation/translation.js');
const { ComponentComposition } = require('../../../manage/models/composition.js');
const { ComponentArtdata } = require('../../../manage/models/artdata.js');
const { exportSvgToImage } = require('./inkscapeProcessor.js');
const { FontCache } = require('./fontCache.js');

async function convertElementToString(element) {
    const builder = new Builder();
    const xml = await builder.buildObject(element.root);
    return xml.replace(/^<\?xml.*?\?>/, '<?xml version="1.0" standalone="yes"?>');
}

async function createArtFileOfPiece(compositions, artdata, gamedata, componentBackOutputDirectory, productionProperties, _) {
    const templateFilesDirectory = compositions.gameCompose["artTemplatesDirectory"];
    if (artdata === null) {
        console.log(`!!! Missing artdata ${gamedata.componentDataBlob["name"]}`);
        return;
    }
    const artFilename = `${artdata["templateFilename"]}.svg`;
    const artFilepath = path.normalize(path.join(productionProperties.inputDirectoryPath, templateFilesDirectory, artFilename));
    if (!await fs.pathExists(artFilepath)) {
        console.log(`!!! Template art file ${artFilepath} does not exist.`);
        return;
    }

    if (!(compositions.componentCompose["type"] in COMPONENT_INFO)) {
        throw new Error(`No image size for ${compositions.componentCompose["type"]}`);
    }
    const component = COMPONENT_INFO[compositions.componentCompose["type"]];

    let contents = null;
    try {
        contents = await fs.readFile(artFilepath, 'utf8');
    } catch (e) {
        console.log(`!!! Template art file ${artFilepath} cannot be parsed. Error: ${e}`);
        return;
    }

    const pieceName = gamedata.pieceData ? gamedata.pieceData["name"] : gamedata.componentBackDataBlob["name"];
    const imageSizePixels = component["DimensionsPixels"];

    try {
        contents = await addOverlays(contents, artdata["overlays"], compositions, gamedata, productionProperties);
        contents = await textReplaceInFile(contents, artdata["textReplacements"], gamedata, productionProperties);
        contents = await updateStylesInFile(contents, artdata["styleUpdates"], gamedata);
        contents = await addNewlines(contents);

        const pieceUniqueHash = gamedata.pieceUniqueBackHash !== '' ? `_${gamedata.pieceUniqueBackHash}` : '';
        const artFileOutputName = `${compositions.componentCompose['name']}${pieceUniqueHash}-${pieceName}`;
        
        const artFileOutputFilepath = await createArtfile(contents, artFileOutputName, imageSizePixels, componentBackOutputDirectory);
        
        await exportSvgToImage(artFileOutputFilepath, imageSizePixels, artFileOutputName, componentBackOutputDirectory);
        console.log(`Produced ${pieceName}.`);
    } catch (error) {
        console.error(`Error producing ${pieceName}: ${error.message}`);
        console.error(error.stack);
    }
}

async function addNewlines(contents) {
    if (!contents) {
        console.error("Warning: contents is null or undefined");
        return "";
    }
    return contents.replace(/NEWLINE/g, "\n");
}

async function createArtfile(contents, artFileOutputName, imageSizePixels, outputDirectory) {
    if (contents === null) {
        throw new Error("Contents cannot be null");
    }

    const parser = new DOMParser();
    const doc = parser.parseFromString(contents, 'image/svg+xml');
    const root = doc.documentElement;

    const scale_factor = 96 / 300;

    root.setAttribute('width', `${imageSizePixels[0] * scale_factor}`);
    root.setAttribute('height', `${imageSizePixels[1] * scale_factor}`);

    root.setAttribute('viewBox', `0 0 ${imageSizePixels[0] * scale_factor} ${imageSizePixels[1] * scale_factor}`);

    const wrapper = doc.createElementNS('http://www.w3.org/2000/svg', 'g');
    wrapper.setAttribute('transform', `scale(${scale_factor})`);

    while (root.firstChild) {
        wrapper.appendChild(root.firstChild);
    }
    root.appendChild(wrapper);

    const serializer = new XMLSerializer();
    contents = serializer.serializeToString(doc);

    const artFileOutputFileName = `${artFileOutputName}.svg`;
    const artFileOutputFilepath = path.join(outputDirectory, artFileOutputFileName);
    await fs.writeFile(artFileOutputFilepath, contents, 'utf8');
    return artFileOutputFilepath;
}

async function addOverlays(contents, overlays, compositions, pieceGamedata, productionProperties) {
    if (contents === null) {
        throw new Error("contents cannot be null");
    }

    if (overlays === null) {
        throw new Error("overlays cannot be null");
    }

    const overlayFilesDirectory = compositions.gameCompose["artInsertsDirectory"];

    for (const overlay of overlays) {
        const isComplex = overlay.isComplex ?? false;
        if (isComplex && productionProperties.isSimple) {
            continue;
        }

        const isDebug = overlay.isDebugInfo ?? false;
        if (isDebug && productionProperties.isPublish) {
            continue;
        }

        const positionX = overlay.positionX ?? 0;
        const positionY = overlay.positionY ?? 0;
        const overlayName = await getScopedValue(overlay, pieceGamedata);
        if (!overlayName) {
            continue;
        }

        const overlaysFilepath = path.resolve(path.join(productionProperties.inputDirectoryPath, overlayFilesDirectory));
        const overlayFilename = `${overlayName}.svg`;
        const overlayFilepath = path.normalize(path.join(overlaysFilepath, overlayFilename));

        if (!await fs.pathExists(overlayFilepath)) {
            console.log(`!!! Overlay ${overlayFilepath} does not exist.`);
            continue;
        }

        contents = await placeOverlay(contents, overlayFilepath, positionX, positionY);
    }
    return contents;
}

async function placeOverlay(contents, overlayFilepath, positionX, positionY) {
    if (!contents) {
        console.error("Warning: contents is null or undefined");
        return "";
    }

    try {
        // Sanitize the main SVG content
        contents = sanitizeSvgContent(contents);
        
        const parser = new DOMParser({
            errorHandler: {
                warning: function(w) { /* Suppress warnings */ },
                error: function(e) { console.error("Error parsing main SVG: ", e); },
                fatalError: function(e) { console.error("Fatal error parsing main SVG: ", e); }
            }
        });
        
        const mainDoc = parser.parseFromString(contents, 'image/svg+xml');
        if (!mainDoc || !mainDoc.documentElement) {
            console.error("Failed to parse main SVG document");
            return contents;
        }
        
        const mainRoot = mainDoc.documentElement;

        let overlayContents = await fs.readFile(overlayFilepath, 'utf8');
        // Sanitize the overlay SVG content
        overlayContents = sanitizeSvgContent(overlayContents);
        
        const overlayDoc = parser.parseFromString(overlayContents, 'image/svg+xml');
        if (!overlayDoc || !overlayDoc.documentElement) {
            console.error(`Failed to parse overlay SVG document: ${overlayFilepath}`);
            return contents;
        }
        
        const overlayRoot = overlayDoc.documentElement;

        const group = mainDoc.createElementNS('http://www.w3.org/2000/svg', 'g');

        const viewBox = mainRoot.getAttribute('viewBox')?.split(/\s+/).map(parseFloat) ?? [];
        let scale_factor = 1.0;
        if (viewBox.length === 4) {
            const [, , width, height] = viewBox;
            const actual_width = parseFloat(mainRoot.getAttribute('width') ?? width);
            scale_factor = actual_width / width;
        }

        group.setAttribute('transform', `translate(${positionX},${positionY}) scale(${1 / scale_factor})`);

        // Copy attributes from overlay root to group, excluding certain attributes
        for (let i = 0; i < overlayRoot.attributes.length; i++) {
            const attr = overlayRoot.attributes[i];
            if (!['width', 'height', 'viewBox', 'transform'].includes(attr.name)) {
                group.setAttribute(attr.name, attr.value);
            }
        }

        // Move all children from overlay root to group
        while (overlayRoot.firstChild) {
            try {
                group.appendChild(overlayRoot.firstChild);
            } catch (e) {
                console.error(`Error appending child: ${e.message}`);
                break;
            }
        }

        mainRoot.appendChild(group);

        const serializer = new XMLSerializer();
        return serializer.serializeToString(mainDoc);
    } catch (error) {
        console.error(`Error in placeOverlay: ${error.message}`);
        return contents;
    }
}

async function textReplaceInFile(contents, textReplacements, gamedata, productionProperties) {
    if (!contents) {
        console.error("Warning: contents is null or undefined");
        return "";
    }

    try {
        for (const textReplacement of textReplacements) {
            const key = `{${textReplacement["key"]}}`;
            let value = await getScopedValue(textReplacement, gamedata);
            if (typeof value === 'string') {
                value = value.replace(/\\\\n/g, "\n");
            }
            value = await processValueFilters(value, textReplacement);

            const isComplex = textReplacement.isComplex ?? false;
            if (isComplex && productionProperties.isSimple) {
                value = "";
            }

            const isDebug = textReplacement.isDebugInfo ?? false;
            if (isDebug && productionProperties.isPublish) {
                value = "";
            }

            if (productionProperties.targetLanguage !== "en" && (textReplacement.isTranslateable ?? false) && value) {
                const translation = getTranslation("./", value, productionProperties.targetLanguage);
                if (translation) {
                    value = translation;
                } else {
                    console.log(`Could not translate ${value}`);
                }
            }
            
            // Do NOT escape XML special characters for text replacements
            // This allows XML tags like <tspan> to be properly injected
            
            contents = contents.replace(key, value || "");
        }
        return contents;
    } catch (error) {
        console.error(`Error in text replacement: ${error.message}`);
        return contents;
    }
}

async function processValueFilters(value, textReplacement) {
    if ("filters" in textReplacement) {
        for (const filter of textReplacement["filters"]) {
            if (filter === "toUpper") {
                value = value.toUpperCase();
            }
        }
    }
    return String(value);
}

async function updateStylesInFile(contents, styleUpdates, pieceGamedata) {
    if (!contents) {
        console.error("Warning: contents is null or undefined");
        return "";
    }

    try {
        // Try to sanitize the SVG content before parsing
        contents = sanitizeSvgContent(contents);
        
        const parser = new DOMParser({
            errorHandler: {
                warning: function(w) { console.warn("Warning: ", w); },
                error: function(e) { console.error("Error: ", e); },
                fatalError: function(e) { console.error("Fatal Error: ", e); }
            }
        });
        
        const doc = parser.parseFromString(contents, 'image/svg+xml');
        
        // Check if doc is valid
        if (!doc || !doc.documentElement) {
            console.error("Error: Invalid document after parsing");
            return contents;
        }
        
        for (const styleUpdate of styleUpdates) {
            const findById = styleUpdate["id"];
            // Use getElementsByTagName and filter by id instead of querySelector
            const elements = doc.getElementsByTagName("*");
            let elementToUpdate = null;
            
            for (let i = 0; i < elements.length; i++) {
                const element = elements[i];
                if (element.getAttribute('id') === findById) {
                    elementToUpdate = element;
                    break;
                }
            }
            
            if (elementToUpdate !== null) {
                const value = await getScopedValue(styleUpdate, pieceGamedata);
                await replaceStyleAttributeForElement(elementToUpdate, "style", styleUpdate["cssValue"], value);
            } else {
                console.log(`Could not find element with id [${findById}].`);
            }
        }
        
        const serializer = new XMLSerializer();
        return serializer.serializeToString(doc);
    } catch (error) {
        console.error(`Error updating styles: ${error.message}`);
        return contents;
    }
}

async function replaceStyleAttributeForElement(element, attribute, key, value) {
    const attributeValue = element.getAttribute(attribute) ?? "";
    let replaceStyleWith = "";
    let found = false;

    const cssKeyValuePairs = attributeValue.split(';');
    for (const cssKeyValuePair of cssKeyValuePairs) {
        const keyAndPair = cssKeyValuePair.split(':');
        if (keyAndPair[0] === key) {
            replaceStyleWith += `${key}:${value};`;
            found = true;
        } else {
            replaceStyleWith += cssKeyValuePair + ';';
        }
    }

    if (!found) {
        replaceStyleWith += `${key}:${value};`;
    }

    if (replaceStyleWith.endsWith(";")) {
        replaceStyleWith = replaceStyleWith.slice(0, -1);
    }
    element.setAttribute(attribute, replaceStyleWith);
}

async function getScopedValue(scopedValue, pieceGameData) {
    if (scopedValue === null) {
        throw new Error("scopedValue cannot be null.");
    }

    const scope = scopedValue["scope"];
    const source = scopedValue["source"];

    let scopeData = null;
    if (scope === "studio") {
        scopeData = pieceGameData.studioDataBlob;
    } else if (scope === "game") {
        scopeData = pieceGameData.gameDataBlob;
    } else if (scope === "component") {
        scopeData = pieceGameData.componentDataBlob;
    } else if (scope === "piece") {
        scopeData = pieceGameData.pieceData ? pieceGameData.pieceData : pieceGameData.componentBackDataBlob;
    } else if (scope === "global") {
        return source;
    } else if (scope === "utility") {
        const utilityFunctions = {
            "git-sha": getCurrentGitSha
        };
        if (!(source in utilityFunctions)) {
            console.log(`Missing function ${source} not found in ${scope} scope.`);
            return source;
        }
        return utilityFunctions[source]();
    }

    if (!(source in scopeData)) {
        console.log(`Missing key ${source} not found in ${scope} scope.`);
        return source;
    }

    return scopeData[source];
}

function getCurrentGitSha() {
    try {
        const { execSync } = require('child_process');
        return execSync('git rev-parse HEAD').toString().trim();
    } catch (error) {
        console.error('Error getting git SHA:', error);
        return 'unknown';
    }
}

// Helper function to sanitize SVG content before parsing
function sanitizeSvgContent(content) {
    if (!content) return "";
    
    // Fix common XML issues
    return content
        // Ensure XML declaration is correct
        .replace(/^<\?xml[^>]*\?>/, '<?xml version="1.0" encoding="UTF-8" standalone="no"?>')
        // Fix self-closing tags
        .replace(/<([a-zA-Z][a-zA-Z0-9]*)\s+([^>]*)\/>/g, '<$1 $2></$1>')
        // Remove any control characters
        .replace(/[\x00-\x09\x0B\x0C\x0E-\x1F\x7F]/g, '')
        // Fix unclosed CDATA sections
        .replace(/<!\[CDATA\[([^\]]*)(?!\]\]>)/g, '<![CDATA[$1]]>')
        // Fix unescaped ampersands
        .replace(/&(?!amp;|lt;|gt;|quot;|apos;|#\d+;|#x[0-9a-fA-F]+;)/g, '&amp;');
}

module.exports = {
    convertElementToString,
    createArtFileOfPiece,
    addNewlines,
    createArtfile,
    addOverlays,
    placeOverlay,
    textReplaceInFile,
    processValueFilters,
    updateStylesInFile,
    replaceStyleAttributeForElement,
    getScopedValue,
    sanitizeSvgContent
};
